cmd/clusterctl/client/cluster/internal/dryrun/client.go

```diff
@@ -0,0 +1,330 @@
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/scheme"
+)
+
+var (
+	localScheme = scheme.Scheme
+)
+
+type identifier struct {
+	gvk schema.GroupVersionKind
+	key client.ObjectKey
+}
+
+type changes struct {
+	created  map[identifier]bool
+	modified map[identifier]client.Object
+	deleted  map[identifier]bool
+}
+
+// Client implements a dry run Client, that is a fake.Client that logs write operations.
+type Client struct {
+	fakeClient client.Client
+	apiReader  client.Reader
+
+	changeTracker *changes
+}
+
+// Modification defined the modification observed on an object.
+type Modification struct {
+	Before *unstructured.Unstructured
+	After  *unstructured.Unstructured
+	Diff   []byte
+}
+
+// Changes defines all the changes detected by the Dryrun execution.
+type Changes struct {
+	Created  []*unstructured.Unstructured
+	Modified []*Modification
+	Deleted  []*unstructured.Unstructured
+}
+
+// NewClient returns a new dry run Client.
+func NewClient(apiReader client.Reader, objs []client.Object) *Client {
+	fakeClient := fake.NewClientBuilder().WithObjects(objs...).WithScheme(localScheme).Build()
+	return &Client{
+		fakeClient: fakeClient,
+		apiReader:  apiReader,
+		changeTracker: &changes{
+			created:  map[identifier]bool{},
+			modified: map[identifier]client.Object{},
+			deleted:  map[identifier]bool{},
+		},
+	}
+}
+
+// Get retrieves an obj for the given object key from the Kubernetes Cluster.
+func (c *Client) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	if err := c.fakeClient.Get(ctx, key, obj); err != nil {
+		// If the object is not found by the fake client, get the object
+		// from the cluster.
+		if apierrors.IsNotFound(err) && c.apiReader != nil {
+			return c.apiReader.Get(ctx, key, obj)
+		}
+		return err
+	}
+	return nil
+}
+
+// List retrieves list of objects for a given namespace and list options.
+// List functions returns the union of the lists from the fake client and the apiReader client.
+// Nb. For objects that exists both in the fake client and in the apiReader, fake client takes precedence.
+func (c *Client) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	var gvk schema.GroupVersionKind
+	if uList, ok := list.(*unstructured.UnstructuredList); ok {
+		gvk = uList.GroupVersionKind()
+	} else {
+		var err error
+		gvk, err = apiutil.GVKForObject(list, c.fakeClient.Scheme())
+		if err != nil {
+			return errors.Wrap(err, "failed to get GVK of target object")
+		}
+	}
+
+	// Fetch lists from both fake client and rhe apiReader and merge the two lists.
+	uFakeList := &unstructured.UnstructuredList{}
+	uReaderList := &unstructured.UnstructuredList{}
+	uFakeList.SetGroupVersionKind(gvk)
+	uReaderList.SetGroupVersionKind(gvk)
+	if err := c.fakeClient.List(ctx, uFakeList, opts...); err != nil {
+		return err
+	}
+	if c.apiReader != nil {
```

Comment:
c.apiReader seems to be always set. Do we need the nil check? (get & list are the only funcs where it's done)

Reply:
Okay sounds good.